app/assets/javascripts/discourse/app/components/uppy-image-uploader.js

```diff
@@ -87,42 +69,21 @@ export default Component.extend(UppyUploadMixin, {
     }
   },
 
-  _openLightbox() {
-    next(() =>
-      $(this.element.querySelector("a.lightbox")).magnificPopup("open")
-    );
+  @on("didRender")
+  _applyLightbox() {
+    next(() => lightbox(this.element, this.siteSettings));
   },
 
-  _applyLightbox() {
-    if (this.imageUrl) {
-      next(() => lightbox(this.element, this.siteSettings));
+  @on("willDestroyElement")
+  _closeOnRemoval() {
+    if (isTesting() && $.magnificPopup?.instance) {
```

Comment:
Any reason why we are only cleaning up on a testing env?

Reply:
So, there's three things to worry about when we have a lightbox.

1. The markup added to the DOM - `.mfp-bg` and `.mfp-wrap` - these are only added if the user opens the lightbox.
2. The event listener added to the lightbox element - `a.lightbox` - this is added when magnific popup is called on an element.
3. The instance stored in memory `$.magnificPopup.instance` - this is created when the user opens the lightbox.

For 1: 
That's what the close method does. It removes those elements from the DOM. We only call it in testing because it's implied that it will be called if the user opens the lightbox. If you open the lightbox, there's only three ways you can get back to the site. The close icon inside the lightbox, pressing <kbd>ESC</kbd> or clicking on the dimmed background behind the image. All three of those will call the close method. So, we don't need to do that when the element is destroyed since it will have already happened via user interaction. We still need to call it in testing because if we don't, the lightbox will stay open even if the test is finished.

For 2: 
Magnific popup does not handle the removal of that listener. It relies on the browser to do it when the `a.lightbox` element is removed from the DOM. In our case, this happens when the user navigates away from the page or when the post is cloaked if it's a topic page. 

For 3:
Sadly Magnific popup has no exposed built-in way to remove that. This PR doesn't change anything in that regard. Calling the close method does not remove that. The last created instance will remain in memory for the rest of the session, even before this PR. The good news, though, is that it only stores the last one. If a user opens a new lightbox, it will replace the old instance.

Reply:
Bear with me for a second because I don't understand all the implications. I just noticed that we do the same thing on each route change: 

https://github.com/discourse/discourse/blob/main/app/assets/javascripts/discourse/app/lib/clean-dom.js#L12-L16

Do you think we still need the extra cleanup on the upload components? 

I'm also seeing an issue locally when running `localhost:4200/tests?filter=upload`: 

<img width="1069" alt="image" src="https://user-images.githubusercontent.com/368961/146603083-df815639-fd52-4342-995d-9b73aaef5a4b.png">

I'm only seeing this on that specific filtered test run (for example, no issues when running all test in command line via `bin/ember-cli --test --launch "Chrome"`, so there might be something else at play in my local environment.

Reply:
oooh nice catch :+1: 

I can reproduce that error as well. It doesn't happen every time, but it does happen. The cause is the removal delay we add here 

https://github.com/discourse/discourse/blob/1472e47aae5bfdfb6fd9abfe89beb186c751f514/app/assets/javascripts/discourse/app/lib/lightbox.js#L29

Which gets used here in a `setTimeout` (L380)

https://github.com/dimsemenov/Magnific-Popup/blob/master/src/js/core.js#L370-L384

Setting that to 0 makes the filtered tests pass every time, but it removes the nice fade-out animation when users close the lightbox.

For the extra cleanup, it's really up to you. As far as I can see, it won't do anything. A user can't exit the lightbox without calling `close` one way or another. The cleanup we do on route change is redundant for users.

So how does this sound?

I'll add the extra cleanup for users (remove the `isTesting` check) and set the removal delay to 0 in tests only.

Reply:
Both removing `isTesting` and setting the delay to 0 for tests sound great, thanks very much!